features/FEATURE_LWIP/lwip-interface/lwip_stack.c

```diff
@@ -651,52 +765,57 @@ nsapi_error_t mbed_lwip_bringup_2(bool dhcp, bool ppp, const char *ip, const cha
     if (ppp) {
        err_t err = ppp_lwip_connect();
        if (err) {
+           lwip_connected = NSAPI_STATUS_DISCONNECTED;
+           if (lwip_client_callback) {
+               lwip_client_callback(lwip_status_cb_handle, NSAPI_EVENT_CONNECTION_STATUS_CHANGE, NSAPI_STATUS_DISCONNECTED);
+           }
            return mbed_lwip_err_remap(err);
        }
     }
 
-    if (!netif_is_link_up(&lwip_netif)) {
-        if (sys_arch_sem_wait(&lwip_netif_linked, 15000) == SYS_ARCH_TIMEOUT) {
-            if (ppp) {
-                ppp_lwip_disconnect();
-            }
-            return NSAPI_ERROR_NO_CONNECTION;
-        }
-    }
-
-    if (!ppp) {
-        netif_set_up(&lwip_netif);
-    }
 
-#if LWIP_DHCP
-    if (stack != IPV6_STACK) {
-        // Connect to the network
-        lwip_dhcp = dhcp;
 
-        if (lwip_dhcp) {
-            err_t err = dhcp_start(&lwip_netif);
-            if (err) {
-                return NSAPI_ERROR_DHCP_FAILURE;
+    if (!netif_is_link_up(&lwip_netif)) {
+        if (lwip_blocking) {
+            if (sys_arch_sem_wait(&lwip_netif_linked, 15000) == SYS_ARCH_TIMEOUT) {
+                if (ppp) {
+                    ppp_lwip_disconnect();
+                }
+                lwip_connected = NSAPI_STATUS_DISCONNECTED;
+                if (lwip_client_callback) {
+                    lwip_client_callback(lwip_status_cb_handle, NSAPI_EVENT_CONNECTION_STATUS_CHANGE, NSAPI_STATUS_DISCONNECTED);
+                }
+                return NSAPI_ERROR_NO_CONNECTION;
             }
         }
+    } else {
+        ret = mbed_set_dhcp(&lwip_netif);
+        if (ret != NSAPI_ERROR_OK) {
+            return ret;
+        }
     }
-#endif
-
+        
     // If doesn't have address
     if (!mbed_lwip_get_ip_addr(true, &lwip_netif)) {
-        if (sys_arch_sem_wait(&lwip_netif_has_any_addr, DHCP_TIMEOUT * 1000) == SYS_ARCH_TIMEOUT) {
-            if (ppp) {
-                ppp_lwip_disconnect();
+        if (lwip_blocking) {
```

Comment:
There should be else clause here that returns if non-blocking. The rest of the function does not do anything, in non-blocking mode, but for clarity ip address should not be checked before link up.